### ISSUE TITLE
[Azure AD OAuth] Single-tenant app registrations not supported

### ISSUE THREAD

**Issue by triantafillos** [Chainlit/chainlit#425]
Currently, the OAuth for Azure AD that Chainlit supports is only for Azure App registrations that are setup with the multi-tenant account type. For App Registrations that are setup as Single-tenant, the OAuth process fails.

<img width="540" alt="Screenshot 2023-09-25 at 10 10 28 PM" src="https://github.com/Chainlit/chainlit/assets/1324795/3a1bd7be-0891-4936-bff2-ea87cad5c156">

<img width="516" alt="Screenshot 2023-09-25 at 6 32 40 PM" src="https://github.com/Chainlit/chainlit/assets/1324795/e4095a14-7953-44e2-b3fc-ba3cd0fb6819">

This happens because the URLs for authorisation and getting the token are different for each setup:

1. For multitenant apps are:

```
https://login.microsoftonline.com/common/oauth2/v2.0/authorize
https://login.microsoftonline.com/common/oauth2/v2.0/token
```

2. For singe tenant apps are:

```
https://login.microsoftonline.com/{tenantId}/oauth2/v2.0/authorize
https://login.microsoftonline.com/{tenantId}/oauth2/v2.0/token
```

Resources:
https://learn.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app
https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-app-types

Additional info:
I have created a commit in my forked repo that via an extra env variable supports both types. you can take a look:
https://github.com/triantafillos/chainlit/commit/f31e7ce66f9de85ec14d46230c60c77fe871982a

**Comment by willydouhard** [comment]
Thanks a lot @triantafillos! Can you take a look @tpatel ?

**Comment by wxp16** [comment]
hello @triantafillos , after setting account type as "Multitenant", I got the following error. ,I did set proxy in my env, are there any special requirements about the proxy setting in `chainlit`, any clue about this ?
```
infos = await self._loop.getaddrinfo(
  File "/usr/lib/python3.10/asyncio/base_events.py", line 863, in getaddrinfo
    return await self.run_in_executor(
  File "/usr/lib/python3.10/asyncio/futures.py", line 285, in __await__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.10/asyncio/tasks.py", line 304, in __wakeup
    future.result()
  File "/usr/lib/python3.10/asyncio/futures.py", line 201, in result
    raise self._exception.with_traceback(self._exception_tb)
  File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/lib/python3.10/socket.py", line 955, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -5] No address associated with hostname
```

**Comment by triantafillos** [comment]
hello @wxp16, while testing Azure AD OAuth with a multi tenant application I faced no issues. I did nothing extra concerning proxies. You only have to setup the env variables that are described in the documentation.

Still, the current issue is about adding support also for single tenant apps. So I think, what you are describing is irrelevant to the current issue.

**Comment by willydouhard** [comment]
> hello @triantafillos , after setting account type as "Multitenant", I got the following error. ,I did set proxy in my env, are there any special requirements about the proxy setting in `chainlit`, any clue about this ?
> 
> ```
> infos = await self._loop.getaddrinfo(
>   File "/usr/lib/python3.10/asyncio/base_events.py", line 863, in getaddrinfo
>     return await self.run_in_executor(
>   File "/usr/lib/python3.10/asyncio/futures.py", line 285, in __await__
>     yield self  # This tells Task to wait for completion.
>   File "/usr/lib/python3.10/asyncio/tasks.py", line 304, in __wakeup
>     future.result()
>   File "/usr/lib/python3.10/asyncio/futures.py", line 201, in result
>     raise self._exception.with_traceback(self._exception_tb)
>   File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
>     result = self.fn(*self.args, **self.kwargs)
>   File "/usr/lib/python3.10/socket.py", line 955, in getaddrinfo
>     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
> socket.gaierror: [Errno -5] No address associated with hostname
> ```

Can you share a minimal reproducible code (without your secrets obviously)? Also your OS and python version?

**Comment by wxp16** [comment]
> > hello @triantafillos , after setting account type as "Multitenant", I got the following error. ,I did set proxy in my env, are there any special requirements about the proxy setting in `chainlit`, any clue about this ?
> > ```
> > infos = await self._loop.getaddrinfo(
> >   File "/usr/lib/python3.10/asyncio/base_events.py", line 863, in getaddrinfo
> >     return await self.run_in_executor(
> >   File "/usr/lib/python3.10/asyncio/futures.py", line 285, in __await__
> >     yield self  # This tells Task to wait for completion.
> >   File "/usr/lib/python3.10/asyncio/tasks.py", line 304, in __wakeup
> >     future.result()
> >   File "/usr/lib/python3.10/asyncio/futures.py", line 201, in result
> >     raise self._exception.with_traceback(self._exception_tb)
> >   File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
> >     result = self.fn(*self.args, **self.kwargs)
> >   File "/usr/lib/python3.10/socket.py", line 955, in getaddrinfo
> >     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
> > socket.gaierror: [Errno -5] No address associated with hostname
> > ```
> 
> Can you share a minimal reproducible code (without your secrets obviously)? Also your OS and python version?

yeah, in window wsl2 env with a poetry managed virtual environment, below is the auth callback method: 
```
@cl.oauth_callback
def oauth_callback(
    provider_id: str,
    token: str,
    raw_user_data: Dict[str, str],
    default_app_user: cl.AppUser,
) -> Optional[cl.AppUser]:
    return default_app_user
```
```

**Comment by willydouhard** [comment]
If you remove the auth callback your chainlit app functions as expected I imagine?

**Comment by wxp16** [comment]
> If you remove the auth callback your chainlit app functions as expected I imagine?

yes. it works as expected.

**Comment by wxp16** [comment]
y

> yes.. don't know how to fix. BTW, for chainlit, how to install from src code, rather than `pip`. I found some bugs, I can test in my local env

**Comment by triantafillos** [comment]
@willydouhard maybe it would be better if @wxp16 opened another issue? The current issue is about Azure AD single tenant apps, while @wxp16 seems to have problems with the proxy conf and now we are taking the conversation towards how to run from source. BTW @wxp16 you can try the [contributing guidelines](https://github.com/Chainlit/chainlit/blob/aae67be728de3ebdf86e0f9ea010ae19eef9fb03/.github/CONTRIBUTING.md)

**Comment by clementsirieix** [comment]
@triantafillos it should be fixed with #443 and we have updated the [documentation](https://docs.chainlit.io/authentication/oauth#azure-active-directory) to cover single-tenant applications. Can you confirm that it's working on your end?

**Comment by triantafillos** [comment]
Thanks a lot @clementsirieix I can confirm that it works as expected!

Two notes from my side regarding #443:

- In case of single-tenant app, the `tenant` param in the `authorize_params` is redundant, as it is passed from the URL of the API call.
- The `os.environ.get("OAUTH_AZURE_AD_ENABLE_SINGLE_TENANT")` is called multiple times, so as an optimization it could be saved as a class attribute
